Faster(?) FP16 gemv kernel

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126297

Differential Revision: [D57369266](https://our.internmc.facebook.com/intern/diff/D57369266/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D57369266/)!

swolchok committed:
[PyTorch] Prototype: faster(?) FP16 gemv kernel



Differential Revision: [D57369266](https://our.internmc.facebook.com/intern/diff/D57369266/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D57369266/)!

[ghstack-poisoned]

malfet commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -215,6 +215,56 @@ static inline float16_t reduce(float16x8_t x) {
+
+}
+
+#ifdef __ARM_FEATURE_FMA
+#define F16_FMA(a, b, c) vfmaq_f16((a), (b), (c))
+#else
+#define F16_FMA(a, b, c) vaddq_f16((a), vmulq_f16((b), (c)))
+#endif
```
Nit (less preprocessor code and more inline functions is better imo)
Also, aren't fma(a, b, c) == a*b + c ? I.e. order of operands in your non-fma implementaiton is wrong
```suggestion
static inline f16_fma(float16x8_t a, float16x8_t b, float16x8_t c) {
#ifdef __ARM_FEATURE_FMA
 return vfmaq_f16(a, b, c);
#else
 return vaddq_f16(a, vmulq_f16(b, c));
#endif
}
```

malfet commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -263,17 +315,23 @@ static void fp16_gemv_trans_fp32_arith(const int m, const int n, const float16_t
       const auto row1 = a + lda * (i + 1);
       const auto row2 = a + lda * (i + 2);
       const auto row3 = a + lda * (i + 3);
+#ifdef __ARM_FEATURE_FMA
+#define F32_FMA(a, b, c) vfmaq_f32((a), (b), (c))
+#else
+#define F32_FMA(a, b, c) vaddq_f32((a), vmulq_f32((b), (c)))
+#endif
```
Same as above, please replace it with inline

swolchok committed:
Update on "[PyTorch] Prototype: faster(?) FP16 gemv kernel"


Differential Revision: [D57369266](https://our.internmc.facebook.com/intern/diff/D57369266/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D57369266/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Prototype: faster(?) FP16 gemv kernel"


Differential Revision: [D57369266](https://our.internmc.facebook.com/intern/diff/D57369266/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D57369266/)!

[ghstack-poisoned]

malfet reviewed (approved):
Looks good to me

malfet commented:
@pytorochbot merge

malfet commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge

malfet commented:
I've tried it with https://github.com/malfet/llm_experiments and fp16 perf on llama2-7b changed from 6.37 to 6.54 tokens/sec
